File: net/JetBrains.Serialization/src/Dataholders/PeEncapContentInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.Cms;
using Org.BouncyCastle.Asn1.X509;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class PeEncapContentInfo : EncapContentInfo
{
  [JsonProperty("ContentType")] protected override TextualInfo ContentType { get; }
  [JsonProperty("ImageDataObjIdInfo")] private ImageDataObjIdInfo _imageDataObjIdInfo;
  [JsonProperty("HashAlgorithmInfo")] private AlgorithmInfo _hashAlgorithmInfo;
  [JsonProperty("ContentHash")] private TextualInfo _contentHash;

  [JsonConstructor]
  public PeEncapContentInfo(TextualInfo contentType, ImageDataObjIdInfo imageDataObjIdInfo,
    AlgorithmInfo hashAlgorithmInfo, TextualInfo contentHash)
  {
    ContentType = contentType;
    _imageDataObjIdInfo = imageDataObjIdInfo;
    _hashAlgorithmInfo = hashAlgorithmInfo;
    _contentHash = contentHash;
  }

  public new static PeEncapContentInfo GetInstance(ContentInfo contentInfo)
  {
    var contentSequence = ((DerSequence)contentInfo.Content).ToArray();
    var algorithmSequence = ((DerSequence)contentSequence[1]).ToArray();

    return new PeEncapContentInfo(
      TextualInfo.GetInstance(contentInfo.ContentType),
      ImageDataObjIdInfo.GetInstance((DerSequence)contentSequence.First()),
      new AlgorithmInfo((AlgorithmIdentifier.GetInstance(algorithmSequence.First()))),
      TextualInfo.GetInstance(algorithmSequence[1]));
  }

  protected override Asn1Encodable GetContentPrimitive()
    =>
      new List<Asn1Encodable?>
      {
        _imageDataObjIdInfo.ToPrimitive(),
        new List<IEncodableInfo>
        {
          _hashAlgorithmInfo,
          _contentHash
        }.ToPrimitiveDerSequence()
      }.ToDerSequence();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/PeEncapContentInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSequence
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.cms.ContentInfo
import org.bouncycastle.asn1.x509.AlgorithmIdentifier

@Serializable
data class PeEncapContentInfo(
  override val contentType: TextualInfo,
  val imageDataObjIdInfo: ImageDataObjIdInfo,
  val hashAlgorithmInfo: AlgorithmInfo,
  val contentHash: TextualInfo
) : EncapContentInfo() {
    companion object {
        fun getInstance(contentInfo: ContentInfo): PeEncapContentInfo =
            (contentInfo.content as DLSequence).let { contentSequence ->
                (contentSequence.getObjectAt(1) as DLSequence).let { algorithmSequence ->
                    PeEncapContentInfo(
                        TextualInfo.getInstance(contentInfo.contentType),
                        ImageDataObjIdInfo.getInstance(contentSequence.first() as DLSequence),
                        AlgorithmInfo(
                            (AlgorithmIdentifier.getInstance(
                                algorithmSequence.first()
                            ))
                        ),
                        TextualInfo.getInstance(algorithmSequence.getObjectAt(1))
                    )
                }
            }
    }

    override fun getContentPrimitive() =
        listOf(
            imageDataObjIdInfo.toPrimitive(),
            listOf(
                hashAlgorithmInfo.toPrimitive(),
                contentHash.toPrimitive()
            ).toDLSequence()
        ).toDLSequence()
}